src/Parallel/Main.hpp

```diff
@@ -355,7 +358,24 @@ Main<Metavariables>::Main(CkArgMsg* msg) {
         ERROR("No default input file name.  Pass --input-file.");
       }
       input_file = parsed_command_line_options["input-file"].as<std::string>();
-      options.parse_file(input_file);
+
+      if constexpr (tmpl::list_contains_v<option_list,
+                                          OptionTags::InputSource>) {
+        auto input = Options::Options_detail::open_file(input_file);
+        const std::string input_yaml{std::istreambuf_iterator(input), {}};
+
+        // Use yaml block syntax to store the entire input source as a string.
+        // This requires adding two spaces to the start of each line of the
+        // input source string
+        std::istringstream input_stream{input_yaml};
+        std::string input_block{"InputSource: |\n"};
+        for (std::string line; std::getline(input_stream, line);) {
+          input_block += "  "s + line + "\n"s;
+        }
+        options.parse(input_yaml + "\n"s + input_block);
```

Comment:
Sorry, but I don't understand what this is doing.  Why are you not just requesting the file contents from the option parser?

Reply:
Items in GlobalCache must be set with an option tag, unless there's a way around this I couldn't figure out. Here I'm adding a new option tag, `InputSource:`, to the file contents before parsing. The value of this option is the contents of the file, but I have to put those contents in block format to preserve newlines, whitespace, etc.

Then, when parsing the options, the contents of the file (i.e., the value of the`InputSource:` option, get stored in the global cache under the `InputSource` tag.

If there's a better way to get the file contents into the global cache, I'd be happy to hear it!

Reply:
Use the `InputSource` option tag you moved in the first commit.  You will have to `tmpl::remove` it from the list of tags passed to the option parser because it is handled specially.

Reply:
@wthrowe I could give this a try, but I'm unclear how your suggestion would work. I could try `tmp:remove`ing the `InputSource` option tag before parsing the options, but then how would I get `InputSource` and its string value into the global cache?  It looks to me like the global cache gets populated on `Main.hpp:442`:

```
Parallel::create_from_options<Metavariables>(options_,
                                                   const_global_cache_tags{}),
```

Here `options_` is a tagged tuple created by the option parser `options` on `Main.hpp:385`. I don't see a way to specially append a tagged tuple with an extra tag/value pair. Can you please go into a little more detail about how I can parse the options with the `InputSource` tag removed and then somehow get the `InputSource` tag and value into the global cache?

Reply:
It's a special case in the option parser.  You can always `Parser::get` (or `Parser::apply`) `InputSource`, even though it is not in the template parameters for the `Parser`.  So leave all the code that pulls things out of the parser the same, but remove `InputSource` from just the tags in the template parameter of the parser itself.